src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -334,30 +340,77 @@ private void failUnpublished(final List<BatchItem> batch, final String reason) {
     }
 
     @Override
-    public Map<String, Long> materializePositions(final String topicId, final SubscriptionBase.InitialPosition position)
+    public Map<String, Long> materializePositions(final EventType eventType, final Subscription subscription)
             throws ServiceUnavailableException {
-        try (final Consumer<String, String> consumer = kafkaFactory.getConsumer()) {
 
-            final org.apache.kafka.common.TopicPartition[] kafkaTPs = consumer
-                    .partitionsFor(topicId)
-                    .stream()
-                    .map(p -> new org.apache.kafka.common.TopicPartition(topicId, p.partition()))
-                    .toArray(org.apache.kafka.common.TopicPartition[]::new);
-            consumer.assign(Arrays.asList(kafkaTPs));
-            if (position == SubscriptionBase.InitialPosition.BEGIN) {
-                consumer.seekToBeginning(kafkaTPs);
-            } else if (position == SubscriptionBase.InitialPosition.END) {
-                consumer.seekToEnd(kafkaTPs);
-            } else {
-                throw new IllegalArgumentException("Bad offset specification " + position + " for topic " + topicId);
+        final SubscriptionBase.InitialPosition position = subscription.getReadFrom();
+
+        if (position == SubscriptionBase.InitialPosition.CURSORS) {
+            final List<SubscriptionCursorWithoutToken> etInitialCursors = subscription.getInitialCursors().stream()
+                    .filter(c -> c.getEventType().equals(eventType.getName()))
+                    .collect(Collectors.toList());
+            return getSubscriptionTopicInitPositions(eventType.getTopic(), etInitialCursors);
+
+        } else {
+            try (final Consumer<String, String> consumer = kafkaFactory.getConsumer()) {
+                final org.apache.kafka.common.TopicPartition[] kafkaTPs = consumer
+                        .partitionsFor(eventType.getTopic())
+                        .stream()
+                        .map(p -> new org.apache.kafka.common.TopicPartition(eventType.getTopic(), p.partition()))
+                        .toArray(org.apache.kafka.common.TopicPartition[]::new);
+                consumer.assign(Arrays.asList(kafkaTPs));
+                if (position == SubscriptionBase.InitialPosition.BEGIN) {
+                    consumer.seekToBeginning(kafkaTPs);
+                } else if (position == SubscriptionBase.InitialPosition.END) {
+                    consumer.seekToEnd(kafkaTPs);
+                } else {
+                    throw new IllegalArgumentException("Bad offset specification " + position + " for topic " +
```

Comment:
Yeah I don't really know why we have this check, I removed it now https://github.com/zalando/nakadi/pull/579/commits/9f92f45884426ac890b7128164a629d1828f3c81